Reworked Bloop connection and Tree View

tgodzik opened the pull request:
Previously, we would only reconnect if the user asked for it and often this would not show up quickly.
Now, we make the reconnection work better by:
- reconnecting as soon as the build server is disconnected (fixes https://github.com/scalameta/metals/issues/1354)
- making reconnection automatic unless the user specifically asks for it not to be via system property `metals.ask-to-reconnect=true`
- moving compilation outside the initial build server connection - longer running compilations would then produce invalid exceptions and break the next connection
- reindexing and compiling the workspace on reconnection (this is needed since workspaces would not have all the classpath updated)

Also in this PR:
-  a new clean-compile command was introduced, which together with the restart bloop command was added to tree view
- tree view now has 4 parts - one for help, one for project view, one for commands and one for running compilations

The reworked tree view would look like this:

![image](https://user-images.githubusercontent.com/3807253/80972250-30f8db80-8e1e-11ea-847b-b74ebc0ba0c3.png)


I am also thinking of:
- adding icons to compiling projects, I've seen it break in some cases
- partly reindexing the workspace - not sure if that is sensible since Bloop has a different directory for each client and when reconnecting we are effectively a different client

tgodzik commented:
@nilskp I changed the TreeView to make it more explicit that the commands are clickable and add two more useful commands. Will this be helpful?

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:
```diff
@@ -96,13 +98,15 @@ final class BloopServers(
         serverStarted
       )
 
+    val finished = Promise[Unit]()
```
This promise is used to determine whether the launcher is finished.

marek1840 replied:
then maybe it could be called: `launched` instead of `finished`?

tgodzik replied:
`finished` is updated when the launcher finishes and not when it launches, so I think this is better.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/BspServers.scala:
```diff
@@ -70,14 +72,21 @@ final class BspServers(
 
+      val finished = Promise[Unit]()
+      Future {
+        process.waitFor()
```
Not 100% sure about this solution, need to test it with another BSP server.

tgodzik replied:
Added a test for it in Bill Suite, seems to work pretty well.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala:
```diff
@@ -32,11 +34,19 @@ class BuildServerConnection private (
     extends Cancelable {
 
   @volatile private var connection = Future.successful(initialConnection)
+  connection.foreach { conn => conn.onConnectionFinished(reconnect) }
```
This uses the promise from the launcher to reconnect when the connection is finished.

olafurpg replied:
I'm not sure I understand the logic here, what is the difference between this and the logic below?

```suggestion
  initialConnection.onConnectionFinished(reconnect)
```

tgodzik replied:
`onConnectionFinished` is coming from the process or the launcher to say that the connection is finished.  `onReconnection` is used to later reindex the workspace.

The alternative is to do it in the MetalsLanguageServer and just do a full reconnect, but then it will be less seamless for the user.

tgodzik replied:
I think this is a bit better than a full reconnect, because we will not lose any requests to the build server.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala:
```diff
@@ -47,6 +57,12 @@ class BuildServerConnection private (
+  def onReconnection(
+      index: BuildServerConnection => Future[Unit]
+  ): Unit = {
+    onReconnection.set(index)
```
We need to reindex the workspace and this is to make sure that we will only try to reindex once we finished the indexing at least once.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala:
```diff
@@ -118,21 +138,45 @@ class BuildServerConnection private (
-        case response if response == Messages.DisconnectedServer.notNow =>
-          notification.dismiss(5, TimeUnit.MINUTES)
-          connection
+    if (config.askToReconnect) {
```
Depends now on the user configuration.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala:
```diff
@@ -118,21 +138,45 @@ class BuildServerConnection private (
+          conn.onConnectionFinished(reconnect)
+          conn
+        }
+        connection.foreach(_ => onReconnection.get()(this))
```
Here we actually specify that we want to reindex.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Compilations.scala:
```diff
@@ -66,6 +66,26 @@ final class Compilations(
     cascadeBatch.cancelCurrentRequest()
   }
 
+  def recompile(): Future[Unit] = {
```
Used to fully recompile a workspace.

marek1840 replied:
if the `fully` part is more important, then maybe it should be called `compileAll` ?

tgodzik replied:
Changed to `recompileAll`

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1483,33 +1487,56 @@ class MetalsLanguageServer(
-      )
-      scribe.error(message, e)
-      BuildChange.Failed
+      .flatMap(compileAllOpen)
```
Moved compiling after recover, since this caused to disconnect the build server if anything happened during the compilation.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1483,33 +1487,56 @@ class MetalsLanguageServer(
+            val result = connectToNewBuildServer(build)
+            build.onReconnection { reconnected =>
+              connectToNewBuildServer(reconnected)
+                .flatMap(compileAllOpen)
```
We also compile, which will effectively just move all the old artifacts to the new BSP directory.

tgodzik commented on metals/src/main/scala/scala/meta/internal/tvp/TreeViewNode.scala:
```diff
@@ -24,18 +24,21 @@ case class TreeViewNode(
       command = MetalsCommand(
         command.title,
-        command.id,
+        "metals." + command.id,
```
Somehow this didn't work with the new command, since the ones declared all have the `metals.` prefix

This might not be needed, but I haven't been able to make the new command work otherwise.

I think we could also just use an echo command for all server commands here.

ckipp01 commented on metals/src/main/scala/scala/meta/internal/metals/MetalsServerConfig.scala:
```diff
@@ -66,6 +66,10 @@ final case class MetalsServerConfig(
       "bloop.embedded.version",
       BuildInfo.bloopVersion
     ),
+    askToReconnect: Boolean = MetalsServerConfig.binaryOption(
```
Instead of introducing this as a server property, can we just introduce it in `InitializationOptions`?

tgodzik replied:
Client editor should never set this. It's only to be set by a user that knows what they are doing. Usually via `.jvmopts` file in the workspace. The default is fine for all clients.

ckipp01 replied:
ahhh ok, that makes sense. 👍

ckipp01 commented on metals/src/main/scala/scala/meta/internal/metals/ServerCommands.scala:
```diff
@@ -76,6 +76,16 @@ object ServerCommands {
+  val CleanCompile = new Command(
+    "compile-clean",
+    "Recompile workspace",
+    """|Recompile of build targets in this workspace.
```
```suggestion
    """|Recompile all build targets in this workspace.
```

tgodzik replied:
Fixed!

ckipp01 commented:
> @ckipp01 Could you maybe take a look at this from the coc-metals perspective? Not sure if a lot of changes will be need or whether I should do something differently.

First try shows the `MetalsBuild` section just fine, but the `MetalsCompile` and is empty and `MetalsProject` doesn't show up :/. Not sure if it's just a small thing or a larger thing that needs to get fixed on the coc-metals side without diving further in.

tgodzik commented:
Seems to be an issue with shutting down currently. The connection might get automatically reestablised :sweat_smile:  Will fix it tomorrow.

tgodzik commented:
> > @ckipp01 Could you maybe take a look at this from the coc-metals perspective? Not sure if a lot of changes will be need or whether I should do something differently.
> 
> First try shows the `MetalsBuild` section just fine, but the `MetalsCompile` and is empty and `MetalsProject` doesn't show up :/. Not sure if it's just a small thing or a larger thing that needs to get fixed on the coc-metals side without diving further in.

Project just changed the name, so maybe we can adjust the names of views too? MetalsCompile should be empty unless we are compiling anything.

tgodzik commented:
> Should "Recompile workspace" be called "Clean compile workspace" instead?

I was thinking about it, but recompile seemed to be clear that we are redoing the whole compilation. I can change don't feel strongly about it

> Should we use the sync icon  for "Import build" for consistency with IntelliJ "Refresh project"? We can use something like a stop icon  for "Restart build server" instead.

Good idea! I wasn't really sure about those two.

> Can we move "Restart build server" to the bottom of the command list in the tree view?

Sure!

> (brainstorming, maybe for a separate PR) I wonder if we should have some form of "Clean import build" which removes `.bloop/` before importing the build  I worry however that we end up having too many troubleshooting commands and users blindly run them all.

Honestly, I have never needed to do that myself. Either clean compile or restart server should help with most situations. There is no need to remove the .json files in .bloop ever, so the only thing we need to do is remove the compile artifacts, which the clean compile should do.

tgodzik commented:
@olafurpg  I updated the icons for import and restart server. I used `debug-stop` for restart, since stop gives us the same icon as for cancel. I also updated the VSCode PR so that if we don't find an icon we try to use the default ones from VS Code.

The image in the description got updated too.

ckipp01 commented on metals/src/main/scala/scala/meta/internal/tvp/MetalsTreeViewProvider.scala:
```diff
@@ -75,6 +76,12 @@ class MetalsTreeViewProvider(
     languageClient.metalsTreeViewDidChange(
       TreeViewDidChangeParams(
         Array(
+          TreeViewNode(
```
Sort of nitpicky, but I'm trying to debug what is going on with `coc-metals` with the branch, and one things I noticed is that this is _technically_ against the spec that we have published. In the `TreeViewNode` class, `viewId` and `lable` aren't marked as `@Nullable`, and in the [spec](https://scalameta.org/metals/docs/editors/tree-view-protocol.html#tree-view-node) it's the same thing. I'm getting an undefined that I'm trying to chase down, and I don't think it's coming from here, but this is the sort of thing that could cause it and confuse someone since there should be somewhat of a guarantee that there is a `label` here. The same for the other two `TreeViewNode`'s below this.

tgodzik replied:
Added TreeViewNode.empty that sets both the id and label.

tgodzik commented on metals/src/main/scala/scala/meta/internal/tvp/MetalsTreeViewProvider.scala:
```diff
@@ -299,18 +317,16 @@ class MetalsTreeViewProvider(
-    val counter = if (size > 0) s" ($size)" else ""
     TreeViewNode(
       Compile,
-      "metals://ongoing-compilations",
```
@ckipp01 Maybe this part is a problem with the `undefined` ? I hoped the node is discovered by id, but maybe we check something else in vim? This is the only big change, the rest is just moving things around

ckipp01 replied:
No, I found the issue. The TVP panel in coc-metals was sort of built with the assumption that apart from the initial view it was never empty. Not that things have been shifted around the the compilations view will return an empty array just isn't accounted for. So I need to adjust the logic for this since I have logic built-in to find the offset of node to update, which is wrong now.

tgodzik replied:
Ok, cool. Thanks for taking a look at it!

marek1840 commented on metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala:
```diff
@@ -47,9 +58,16 @@ class BuildServerConnection private (
   /** Run build/shutdown procedure */
   def shutdown(): Future[Unit] = connection.map { conn =>
     try {
+      isShuttingDown.set(true)
```
by using `compareAndSet(false, true` you could guarantee here that the method will be called exactly once, hence making it idempotent.

tgodzik replied:
Good idea! Fixed it!

marek1840 commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1483,33 +1488,56 @@ class MetalsLanguageServer(
-      }
-      _ = {
-        treeView.init()
+    def compileAllOpen: BuildChange => Future[BuildChange] = {
```
could be `compileOpenFiles` at first I was not 100% if it meant open files.

tgodzik replied:
changed to `compileAllOpenFiles`
